### PR TITLE
feat(core): added the ability to use Imagick as the image processor

### PR DESCRIPTION
**Pull request by jdalsem**

### CODE REVIEW

File: elgg-config/settings.example.php [hunk 1]
```diff
@@ -329,3 +329,14 @@
...
 $CONFIG->allow_phpinfo = false;
+
+/**
+ * Configure image processor
+ *
+ * This setting can be used to select a different image processor. By default the GD library is used.
+ * Currently only 'imagick' is supported as a different configuration.
+ * For Imagick the 'imagick' extension is required.
+ *
+ * @global bool $CONFIG->image_processor
```
Comment (jeabakker) [inline elgg-config/settings.example.php]: not a `bool`, but `string`

**Review by jeabakker — CHANGES_REQUESTED** [review]

File: engine/classes/Elgg/Di/ServiceProvider.php [hunk 2]
```diff
@@ -282,7 +282,18 @@ public function __construct(Config $config) {
 		$this->setClassName('input', \Elgg\Http\Input::class);
 
 		$this->setFactory('imageService', function(ServiceProvider $c) {
-			$imagine = new \Imagine\Gd\Imagine();
+			$image_processor = _elgg_config()->image_processor;
+			switch ($image_processor) {
+				case 'imagick':
+					if (extension_loaded('imagick')) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Di/ServiceProvider.php]: And what if not?

**Comment by hypeJunction** [comment]
I think it would be cleaner to add a setter to image service and add public API to replace it

File: engine/classes/Elgg/Di/ServiceProvider.php (hunk 2 above)
Comment (jeabakker) [reply engine/classes/Elgg/Di/ServiceProvider.php]: it'll fallback to GD

**Comment by jeabakker** [comment]
tested and it works

File: engine/classes/Elgg/Di/ServiceProvider.php (hunk 2 above)
Comment (hypeJunction) [reply engine/classes/Elgg/Di/ServiceProvider.php]: I don't see how

Comment (jeabakker) [reply engine/classes/Elgg/Di/ServiceProvider.php]: the break is only inside the if, otherwise it continues to the default case

Comment (hypeJunction) [reply engine/classes/Elgg/Di/ServiceProvider.php]: Ah, not very readable if you ask me

Comment (jeabakker) [reply engine/classes/Elgg/Di/ServiceProvider.php]: that\s how PHP switches work ;)

Comment (hypeJunction) [reply engine/classes/Elgg/Di/ServiceProvider.php]: Adding breaks whereever?

Comment (jeabakker) [reply engine/classes/Elgg/Di/ServiceProvider.php]: no the not having breaks and thus continuing to the next case

Comment (hypeJunction) [reply engine/classes/Elgg/Di/ServiceProvider.php]: But there is a break wrapped in an if statement

Comment (jeabakker) [reply engine/classes/Elgg/Di/ServiceProvider.php]: FYI https://secure.php.net/manual/en/control-structures.break.php

**Comment by hypeJunction** [comment]
I don't like how this hangs on conditions. What if I want to use gmagick? Or write a custom driver? Or a new better library is released in the future? Make this OOP and it will last a lifetime without the need for adding more switches and updating config docs.

File: engine/classes/Elgg/Di/ServiceProvider.php (hunk 2 above)
Comment (hypeJunction) [reply engine/classes/Elgg/Di/ServiceProvider.php]: I know how switches work, I am saying this is not readable

**Comment by hypeJunction** [comment]
Also DI factory is not an ideal place for this logic, as it may lead to race conditions etc.

File: engine/classes/Elgg/Di/ServiceProvider.php [hunk 3]
```diff
@@ -282,7 +282,18 @@ public function __construct(Config $config) {
 		$this->setClassName('input', \Elgg\Http\Input::class);
 
 		$this->setFactory('imageService', function(ServiceProvider $c) {
-			$imagine = new \Imagine\Gd\Imagine();
+			$image_processor = _elgg_config()->image_processor;
```
Comment (jeabakker) [inline engine/classes/Elgg/Di/ServiceProvider.php]: use `$c->config->image_processor`

**Review by jeabakker — CHANGES_REQUESTED** [review]

**Review by jeabakker — APPROVED** [review]

**Comment by jdalsem** [comment]
> I don't like how this hangs on conditions. What if I want to use gmagick? Or write a custom driver? Or a new better library is released in the future?

you can do this easily with a plugin (see https://github.com/ColdTrick/imagick for example)

> Also DI factory is not an ideal place for this logic, as it may lead to race conditions etc.

A factory is a fine place for it

> I think it would be cleaner to add a setter to image service and add public API to replace it

Feel free to PR. Config flag will be needed to be able to configure this.

**Comment by hypeJunction** [comment]
> A factory is a fine place for it

I will agree when I see tests

> Config flag will be needed to be able to configure this.

Sure, let's drop the plugin hooks system and move everything to config

**Comment by jdalsem** [comment]
> Sure, let's drop the plugin hooks system and move everything to config

which image processor to use is something a sysadmin will most likely configure. I do not think it makes sense to determine per case if GD or Imagick or something else is used

**Comment by hypeJunction** [comment]
Same goes for email transport, but config is crowded already as it is. I think giving the plugin this freedom is simpler than adding untestable conditionals to the heart of Elgg. But whatever, it's merged already.

**Comment by mrclay** [comment]
This should be reverted. We should not make software by steamrolling over over objections. This did not warrant urgent need to merge.

**Comment by mrclay** [comment]
> you can do this easily with a plugin (see https://github.com/ColdTrick/imagick for example)

@Elgg/core When the need arises to call explicitly-marked private API functions in a plugin (and this does happen), please add a comment like `// DO NOT CALL PRIVATE API. This may break with any future Elgg update.` or consider not releasing the plugin. Many of us have been guilty of this.

> A factory is a fine place for it

This isn't an argument. In this particular case, since this is related to PHP config, I think $CONFIG is not a bad choice for this. Image PHP libraries don't seem to be so prevalent that we need the complete flexibility of injecting an object, but we need to (elsewhere) come up with general guidelines for how/when to let plugins replace our dependencies.

File: engine/classes/Elgg/Di/ServiceProvider.php [hunk 4]
```diff
@@ -282,7 +282,18 @@ public function __construct(Config $config) {
...
 
 		$this->setFactory('imageService', function(ServiceProvider $c) {
-			$imagine = new \Imagine\Gd\Imagine();
+			switch ($c->config->image_processor) {
+				case 'imagick':
+					if (extension_loaded('imagick')) {
+						$imagine = new \Imagine\Imagick\Imagine();
+						break;
+					}
+				default:
```
Comment (mrclay) [inline engine/classes/Elgg/Di/ServiceProvider.php]: If it's set and not supported, we should complain loudly.

Comment (hypeJunction) [reply engine/classes/Elgg/Di/ServiceProvider.php]: We could as well default to imagick if it's supported. Not sure if GD requirements checked anywhere in the installer

**Comment by hypeJunction** [comment]
My preference is plugin hooks or new public API over config settings that contain string of texts and put the application in a volatile state based on ifs and maybes. Injecting an object is testable by creating a mock image driver, testing for imagick string in config and unpredictable behaviour in service factory is a hassle. As for the plugins, we should either expose a pre-boot event to replace services (maybe look for boot.php in plugin root) and do other naughty things, or decouple hooks service from the DI container (see session PR) and trigger a hook before default core factory is called allowing plugins to build their own services.
There is too much coupling in DI currently, I would like to see the session, db, logger and hooks out of services, living their independent lives. Other non-sense services can be exposed to plugins to mess with. That's my opinion, but I am not going to do anything about it so take or leave it. I have resolved to not write anymore open source code, so my involvement is limited to offering advise, if anyone cares to listen to what my outtake from working on tests had been.

File: engine/classes/Elgg/Di/ServiceProvider.php (hunk 4 above)
Comment (mrclay) [reply engine/classes/Elgg/Di/ServiceProvider.php]: I wonder if Imagine has a factory that just picks the best available.

Comment (hypeJunction) [reply engine/classes/Elgg/Di/ServiceProvider.php]: I don't think there is a definition of "best" to work with here. Imagick seems to be the most performant

**Comment by mrclay** [comment]
> I would like to see the session, db, logger and hooks out of services, living their independent lives

Can you explain what you mean here in a new ticket?

**Comment by hypeJunction** [comment]
#11220

**Pull request merged**